Add aws-k8s-1.21-nvidia variant

arnaldo2792 opened the pull request:
**Issue number:**
N / A

**Description of changes:**

Draft PR for aws-k8s-1.21-nvidia variant.

This PR adds the `aws-k8s-1.21-nvidia` variant, along with all the building blocks required to support NVIDIA GPUs in future releases. This variant includes all the configurations and software required to leverage GPUs in containerized workloads. In this variant, the default containerd runtime's name is `nvidia`, and uses the `shimpei` helper program as the OCI runtime binary.

### kmod-5.10-nvidia

The drivers are subpackages of `kmod-5.10-nvidia`, since we don't want to have a spec file per driver type per driver version. The spec file for `kmod-5.10-nvidia` will hold the drivers that are compatible with the 5.10 kernel. New spec files should be added for newer kernel and driver versions. The `kmod-5.10-nvidia` package provides a tmpfilesd conf file to create the lib modules directory where the kernel modules will be created.

Each subpackage installs the libraries and binaries underneath `%{_cross_bindir}/nvidia/<type>` and `%{_cross_libdir}/nvidia/<type>` respectively to prevent collisions while building the subpackages. Kernel module objects are installed in `%{_cross_datadir}/nvidia/<type>/modules`, so that `driverdog` can link them at runtime.

Each subpackage provides a drop-in configuration file for containerd, that sets the `NVIDIA_PATH` environment variable. This environment variable must be set to the directory that contains the NVIDIA userland tools, which will be mounted on the containers by `libnvidia-container`. The environment variable is set for containerd, since `libnvidia-container` is called by the runtime to set up the containers.

```
variants: add aws-k8s-1.21-nvidia
actions: add matrix `fetch-upstream` variable
containerd: add configuration for NVIDIA runtime
packages: add NVIDIA tesla 470 driver
```

Remaining tasks:
- [x] Use `oci-add-hooks` instead of `nvidia-container-runtime`
- [x] Add documentation for driverdog
- [x] Check that Node Feature Discovery, GPU feature Discovery, and DCGM work as expected
- [x] Check aarch64 build works 
- [x] Add documentation for variant

**Testing done:**
I launched p3/g4dn/g5g instances and deployed the `vectoradd-cuda` sample app, I had access to `nvidia-smi` and I ran the sample app:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cuda-vectoradd
spec:
  restartPolicy: OnFailure
  containers:
  - name: cuda-vectoradd
    image: nvidia/samples:vectoradd-cuda11.2.1
    resources:
      limits:
         nvidia.com/gpu: 1
    command: ['sh', '-c', 'sleep infinity']

```

```shell
root@cuda-vectoradd:/# nvidia-smi
Tue Nov  2 01:59:52 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   34C    P0    22W / 300W |      0MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

root@cuda-vectoradd:/# /tmp/sample
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
containerd: add configuration for NVIDIA runtime

The *-nvidia variants use a different runtime instead of runc to inject
prestart hooks using shimpei

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 committed:
variants: add aws-k8s-1.21-nvidia

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Forced push includes:

- Cleaned `containerd-config-toml_k8s`, there were some leftover configurations
- Remove `nvidia-container-runtime` and use `shimpei` instead
- Modify `containerd.service` to read `/etc/containerd/containerd.env` file for additional env variables
- Comments in `kmod-5.10-nvidia` spec file that explains the included files
- Use `shimpei` + `oci-add-hooks` as the `nvidia` runtime for `aws-k8s-1.21-nvidia`
- Documentation for `driverdog`
- Cleaned `driverdog`, there were some leftover errors

bcressey commented on packages/binutils/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+
+[[package.metadata.build-package.external-files]]
+url = "https://ftp.gnu.org/gnu/binutils/binutils-2.37.tar.xz"
+sha512 = "5c11aeef6935860a6819ed3a3c93371f052e52b4bdc5033da36037c1544d013b7f12cb8d561ec954fe7469a68f1b66f1a3cd53d5a3af7293635a90d69edd15e7"
```
This should be binutils 2.35.2, to match the version we use the the SDK, which is what we use to link the kernel. It's a bit annoying that we have to align them since it'll be easy to forget, or update by accident. Might be a good idea to add a test to the spec that `%__ld -v` and `%{version}` are the same.

bcressey commented on packages/binutils/binutils.spec:
```diff
@@ -0,0 +1,70 @@
+  --disable-gold \
+  --disable-gdb \
+  --with-system-zlib \
+  --with-system-readline \
```
We're not pulling in our `readline-devel` - does this have an effect?

arnaldo2792 replied:
I read the `configure`, and by providing the `with-system-readline` the `libreadline.a` file won't be compiled. I searched for references to the development files provided by `readline` in binutils but I didn't find any (I looked for `<readline/*whatever>` I don't have the regex, but you get the gist), so I think overall `with-system-readline` doesn't have any effect in binutils itself, but it does save us an extra compilation.

bcressey commented on packages/binutils/binutils.spec:
```diff
@@ -0,0 +1,70 @@
+%{_cross_attribution_file}
+%{_cross_bindir}/ld
+%{_cross_bindir}/strip
+%{_cross_libdir}/bfd-plugins/libdep.so
```
Do we need this? Can we avoid building it?

arnaldo2792 replied:
It turns out that this library isn't compiled in 2.35.2 :D, I removed it.

bcressey commented on packages/binutils/binutils.spec:
```diff
@@ -0,0 +1,70 @@
+%exclude %{_cross_bindir}/size
+%exclude %{_cross_bindir}/strings
+
+
```
nit: extra newline

bcressey commented on packages/containerd/containerd.spec:
```diff
@@ -7,6 +7,7 @@
 %global gitrev 72cec4be58a9eb6b2910f5d10f1c01ca47d231c0
 
 %global _dwz_low_mem_die_limit 0
+%global _is_nvidia_variant %(if echo %{_cross_variant} | grep -Fqw "nvidia"; then echo 1; else echo 0; fi)
```
I'd very much like to avoid making containerd variant-specific - right now `os` is our only package that's marked as "variant-sensitive". What we've done previously is just to include all the containerd templates.

samuelkarp replied:
@bcressey can correct me if I'm wrong here, but my understanding is that this would break build-caching between variants

bcressey commented on packages/glibc/glibc.spec:
```diff
@@ -93,9 +94,11 @@ make %{?_smp_mflags} -O -r
+install -d %{buildroot}%{_cross_factorydir}/etc/
+
 install -p -m 0644 %{S:1} %{buildroot}%{_cross_tmpfilesdir}/glibc.conf
+install -p -m 0644 %{S:2} %{buildroot}%{_cross_factorydir}/etc/ld.so.conf
```
style nit:
```suggestion
install -d %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}

install -p -m 0644 %{S:1} %{buildroot}%{_cross_tmpfilesdir}/glibc.conf
install -p -m 0644 %{S:2} %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}ld.so.conf
```

bcressey commented on packages/kernel-5.10/config-bottlerocket:
```diff
@@ -73,3 +73,6 @@ CONFIG_KERNEL_ZSTD=y
 CONFIG_DECOMPRESS_ZSTD=y
+
+# Add support for IPMI drivers
+CONFIG_IPMI_HANDLER=y
```
Can we make this a module? If we need it automatically loaded for the nvidia modules, we could use a `modules-load.d` drop-in.

bcressey commented on packages/kmod-5.10-nvidia/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/makedumpfile/makedumpfile/archive/1.6.9/makedumpfile-1.6.9.tar.gz"
+sha512 = "9982985498ae641d390c3b87d92aecd263a502f1a4a9e96e145d86d8778b9e778e4ee98034ef4dbe12ed5586c57278917ea5f3c9ae2989ad1ac051215e03b3d9"
```
I assume this is just a placeholder - can omit the `external-files` block.

arnaldo2792 commented:
Forced push includes:

- Update Github Actions to build the variant
- Add missing symlinks for NVIDIA libraries, used by workloads like the GPU Feature Discovery
- Add NVIDIA tesla 470 license to package
- Add `/etc/kubernetes/node-feature-discovery/features.d` for GPU Feature Discovery to work
- Remove unnecessary patch in `libnvidia-container`
- Create the nvidia-container-toolkit configuration in `/etc/nvidia-container-runtime/config.toml`, otherwise it isn't read, even though it was specified with the `-config` flag

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tesla-470-build-config.toml:
```diff
@@ -0,0 +1,20 @@
+lib-modules-path = "kernel/drivers/extra/video/nvidia/tesla"
+
+objects-source = "/usr/share/nvidia/tesla/470.57.02/modules"
```
Any file that references the full version should be a template (`*.in`) that we use `sed` to rewrite at build time - the way we handle some of the mount units in the "release" package.

bcressey replied:
It seems like this config format is missing a top-level collection of some kind.

We might have multiple sets of modules we're trying to link, each with an object path and an output path.

Maybe a hash map, and `driverdog` would take a key as an argument to decide which module set to link?

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tesla-470-tmpfiles.conf:
```diff
@@ -0,0 +1,4 @@
+C /etc/drivers/driver-config.toml - - - - /usr/share/factory/nvidia/tesla/470.57.02/build-config.toml
+C /etc/containerd/containerd.env - - - - /usr/share/factory/nvidia/tesla/470.57.02/nvidia-path.env
```
It might be better to call this `nvidia.env` and use a conditional include in `containerd.service` - that way we're not claiming the "generic" `containerd.env` file which could be useful for other functionality.

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tmpfiles.conf:
```diff
@@ -0,0 +1 @@
+D /lib/modules/5.10.59/kernel/drivers/extra/video/nvidia/tesla 0755 root root -
```
This also needs to be templated so it doesn't drift away from the actual kernel 5.10 version we're using.

bcressey commented on packages/kubernetes-1.21/kubernetes-tmpfiles.conf:
```diff
@@ -3,3 +3,5 @@ L /etc/kubernetes/manifests - - - - static-pods
 L /etc/kubernetes/secrets-store-csi-providers  - - - - /var/lib/kubelet/providers/secrets-store
 r! /var/lib/kubelet/cpu_manager_state
+d /var/lib/kubelet/node-feature-discovery/features.d - - - -
+L /etc/kubernetes/node-feature-discovery/features.d - - - - /var/lib/kubelet/node-feature-discovery/features.d
```
It'd be good to make this change for all Kubernetes variants.

bcressey commented on packages/libnvidia-container/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+sha512 = "b338d0930d444ac9f411993f06cefdd7b5d12b01aebc8fdf027304d42b9943e038a4067b5ecc0c632c5de32cf4636be5edce4c8ffe17af0a9512202b9bffe688"
+[[package.metadata.build-packages.external-files]]
+url = "https://downloads.sourceforge.net/project/libtirpc/libtirpc/1.1.4/libtirpc-1.1.4.tar.bz2"
+sha512 = "392f391f9fc1bd68d81dc44e4058831a64b32790b5c8c37338b0ab416fad2ae4d16389e632596734dba09780347918cc65c6f134e0c1afd09e81ec250785ed23"
```
We shouldn't need this old version of libtirpc.

bcressey commented on packages/libnvidia-container/libnvidia-container.spec:
```diff
@@ -0,0 +1,76 @@
+%{summary}.
+
+%package devel
+Summary: NVIDIA contanier runtime library development files
```
typo in `container`, but also our standard `devel` summary is "Files for development using the [main package summary]".

```suggestion
Summary: Files for development using the NVIDIA container runtime library
```

bcressey commented on packages/libnvidia-container/libnvidia-container.spec:
```diff
@@ -0,0 +1,76 @@
+%license NOTICE LICENSE COPYING COPYING.LESSER
+%{_cross_libdir}/*.so.*
+%{_cross_attribution_file}
+%{_cross_bindir}/*
```
How many binaries and libraries are included? Can we list them out if it's only one or two?

bcressey commented on sources/logdog/conf/logdog.aws-k8s-1.21-nvidia.conf:
```diff
@@ -0,0 +1 @@
+aws-k8s.conf
```
Should this also check the status of the persistence daemon?

arnaldo2792 replied:
As we discussed offline, I removed the `nvidia-persistenced` daemon from this PR, since it will try to look for devices under `/dev`, which take a while to be created by the kernel. We need to figure out a way to make the daemon wait or be triggered once the devices are created (a.k.a `Wants=nvidia0.device`).

bcressey commented on packages/oci-add-hooks/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+glibc = { path = "../glibc" }
+
+[dependencies]
+nvidia-container-toolkit = { path = "../nvidia-container-toolkit" }
```
This dependency seems like it's the wrong way around - `oci-add-hooks` is a generic helper, and we'd ship `nvidia-hook.json` in some other package that required it.

bcressey commented on packages/oci-add-hooks/oci-add-hooks.spec:
```diff
@@ -0,0 +1,59 @@
+%global _is_nvidia_variant %(if echo %{_cross_variant} | grep -Fqw "nvidia"; then echo 1; else echo 0; fi)
```
This package shouldn't be variant specific. If we reverse the dependency then neither package needs to be variant-specific since we just won't build or include `nvidia-container-toolkit` in the non-nvidia variants.

bcressey commented on packages/oci-add-hooks/oci-add-hooks.spec:
```diff
@@ -0,0 +1,59 @@
+tar -C GOPATH/src/github.com/bitly/go-simplejson -xzf %{SOURCE1} --strip 1
+cp GOPATH/src/github.com/bitly/go-simplejson/LICENSE LICENSE.go-simplejson
+tar -C GOPATH/src/github.com/joeshaw/json-lossless -xzf %{SOURCE2} --strip 1
+cp GOPATH/src/github.com/joeshaw/json-lossless/LICENSE LICENSE.json-lossless
```
Why do we have to manage the vendored dependencies for this package but not for our other third-party Go packages? It'd be worth a comment to explain.

bcressey commented on packages/os/Cargo.toml:
```diff
@@ -39,3 +41,4 @@ glibc = { path = "../glibc" }
 # kexec-tools and makedumpfile required for prairiedog functionality
 # kexec-tools = { path = "../kexec-tools" }
 # makedumpfile = { path = "../makedumpfile" }
+# binutils = { path = "../bintuils" }
```
```suggestion
# binutils = { path = "../binutils" }
```

bcressey commented on packages/os/driver-tmpfiles.conf:
```diff
@@ -0,0 +1 @@
+d /run/cache/thar-be-updates 0755 root root -
```
What's this for?

bcressey commented on packages/os/os.spec:
```diff
@@ -242,6 +249,18 @@ Summary: Setting generator for populating admin container user-data from IMDS.
+Requires: %{_cross_os}binutils
+
+%package -n %{_cross_os}shimpei
+Summary: OCI-compatible shim around oci-add-hooks
```
```suggestion
Summary: OCI-compatible shim around oci-add-hooks
Requires: %{_cross_os}oci-add-hooks
```

bcressey commented on packages/os/driverdog.service:
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Load additional drivers
```
Is this more accurate?

```suggestion
Description=Link and load additional drivers
```

bcressey commented on packages/release/prepare-local.service:
```diff
@@ -43,6 +43,13 @@ ExecStart=/usr/bin/mkdir -p \
+ExecStart=/usr/bin/mkdir -p \
+    ${LOCAL_DIR}/var/lib/kernel-modules/upper \
+    ${LOCAL_DIR}/var/lib/kernel-modules/work
+
```
As a (tiny) optimization, it might be worth folding these into the `kernel-devel` commands above.

bcressey commented on packages/systemd/systemd.spec:
```diff
@@ -11,6 +11,7 @@ Source0: https://github.com/systemd/systemd-stable/archive/v%{version}/systemd-s
 Source1: var-run-tmpfiles.conf
 Source2: systemd-modules-load.conf
 Source3: journald.conf
+Source4: ldconfig-service.conf
```
Can we move this drop-in to the package that more precisely requires it?

I don't think it hurts us too much to delay `ldconfig` like this, but it does mean we're running even more programs before the cache is available.

bcressey commented on sources/driverdog/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+toml = "0.5.8"
+
+
+
```
nit: extra newlines

bcressey commented on packages/os/driverdog.service:
```diff
@@ -0,0 +1,14 @@
+Description=Load additional drivers
+RefuseManualStart=true
+RefuseManualStop=true
+After=configured.target
```
```suggestion
After=preconfigured.target
Wants=preconfigured.target
```

arnaldo2792 commented:
Forced push includes:
- Removed leftover packages from kmod-5.10-nvidia Cargo.toml
- Removed `driver-tmpfiles.conf` leftover file from os package
- Cleaned up kmod-5.10-nvidia spec file and driverdog Cargo.toml

bcressey commented on packages/kmod-5.10-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=NVIDIA persistend service
```
```suggestion
Description=NVIDIA persistenced service
```

bcressey commented on packages/kmod-5.10-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=NVIDIA persistend service
+Before=containerd.service
```
```suggestion
Before=containerd.service
After=configured.target
Wants=configured.target
```

bcressey replied:
Do we actually need "Before"? What's the worst case if the daemon somehow didn't start before containerd managed to launch a container that used GPU devices?

arnaldo2792 replied:
Nothing terrible will happen, it could be started after containerd, this daemon will only keep the GPUs initialized.

bcressey commented on packages/kmod-5.10-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,11 @@
+
+[Service]
+Type=forking
+ExecStopPost=/bin/rm -rf /var/run/nvidia-persistenced
```
`/run` is a tmpfs - what size is the state that is persisted there?

arnaldo2792 replied:
The directory only contains a socket and a file with the daemon pid

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+    #[serde(rename(deserialize = "objects-source"))]
+    objects_source: String,
+    #[serde(rename(deserialize = "driver-type"))]
+    driver_type: String,
```
Where is `driver_type` used?

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+    command(DEPMOD_BIN_PATH, &args)?;
+    info!("Updated modules dependencies");
+
+    // Load kernel modules
+    let mut args = vec!["-a".to_string()];
+    args.append(&mut kernel_modules);
+    command(MODPROBE_BIN_PATH, &args)?;
+    info!("Loaded kernel modules");
```
Linking the modules makes sense to me, but I'm not sure about running `modprobe`, and even `depmod` is a gray area. My misgiving is that we're basically managing global state but doing it from a place where we can't know it's safe - there might be another `driverdog` process running, or we might've loaded the modules previously and that will fail, or `depmod` might encounter some transient files from a kmod build in progress and bomb out.

When we look at adding support for multiple versions of drivers, a workflow like this might make sense:
* spawn `driverdog` and have it link all modules at once, writing to versioned output directories
* assuming the modules can't all be installed under `/lib/modules` because of filename conflicts, we could add a symlink from `/lib/modules/.../kernel/drivers/extra/video/nvidia -> /var/lib/nvidia/driver`, and another symlink from `/var/lib/nvidia/driver -> tesla`.
* every time we update that second symlink, we'd re-run `depmod`
* unloading and reloading a set of nvidia drivers seems fraught with complications, but for the initial install then it could be just `modprobe nvidia`. Or we might regenerate a `modprobe.d` config file and then call `modprobe nvidia`.

Whether or not that's the direction we end up taking, I feel like putting the functionality here is inviting us to expand the charter of this helper program. The alternative is that we run `driverdog`, `depmod`, and `modprobe` in quick succession which seems equally easy to do today, and doesn't lock us into an architectural direction.

bcressey commented on packages/nvidia-container-toolkit/nvidia-container-toolkit-config.toml:
```diff
@@ -0,0 +1,8 @@
+disable-require = false
```
Add a comment to explain what this means? Is it not the default?

bcressey commented on packages/nvidia-container-toolkit/nvidia-container-toolkit-config.toml:
```diff
@@ -0,0 +1,8 @@
+root = "/"
+path = "/usr/bin/nvidia-container-cli"
+environment = []
+debug = "/var/log/nvidia-container-runtime-hook.log"
```
Can we log to the journal instead, or make sure debug logging is off by default? I'm wary of logging to a file that we never truncate or rotate.

arnaldo2792 replied:
No :cry: , I removed the line, since the file is in `/etc` customers can change it for some reason we need debug information.

bcressey commented on packages/nvidia-container-toolkit/nvidia-container-toolkit-config.toml:
```diff
@@ -0,0 +1,8 @@
+path = "/usr/bin/nvidia-container-cli"
+environment = []
+debug = "/var/log/nvidia-container-runtime-hook.log"
+ldconfig = "@/sbin/ldconfig"
```
This is weird syntax, and it's also strange for `nvidia-container-cli` to be updating the ldconfig cache. Can we either omit this so it doesn't try to `ldconfig`, or give it `/bin/true` instead?

arnaldo2792 replied:
`nvidia-container-cli` won't be updating the host's cache, but rather the container's [ld cache](https://github.com/NVIDIA/libnvidia-container/blob/master/src/nvc_ldcache.c#L362), using the `ldconfg` binary of the host. I tried to use the default value which points to the binary in the container, but containers like the GPU Device Plugin fail, maybe because the `ldconfig` binary in the container is somewhat old. Other containers didn't have the same problem. I think it is safe to keep the value I'm using since as I said, `libnvidia-container` won't try to update the host's cache.

samuelkarp commented on packages/release/lib-modules.mount.in:
```diff
@@ -0,0 +1,15 @@
+What=overlay
+Where=PREFIX/lib/modules
+Type=overlay
+Options=noatime,nosuid,nodev,lowerdir=/lib/modules,upperdir=/var/lib/kernel-modules/upper,workdir=/var/lib/kernel-modules/work,context=system_u:object_r:state_t:s0
```
@bcressey should we add anything specific to the SELinux policy covering files in `/var/lib/kernel-modules`?

bcressey replied:
Yes, this would be good to do for all the host directories we create for overlayfs mechanics.

samuelkarp commented on sources/driverdog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+/*!
+  driverdog is a tool to link kernel modules at runtime. It uses a toml configuration file with the following shape:
```
Why the extra indent?  That shows up in the readme too.

samuelkarp commented on sources/shimpei/src/main.rs:
```diff
@@ -0,0 +1,87 @@
+/// Path to runc binary
+const RUNC_BIN_PATH: &str = "/usr/bin/runc";
+
+/// Path to hooks definition, provided by runtime "vendor"
+const HOOK_CONFIG_PATH: &str = "/usr/share/oci-add-hooks/hook.json";
```
Since this is actually vendor-specific, it might be worthwhile to hard-code the (only) vendor path we have today: `/usr/share/oci-add-hooks/nvidia.json`.  When we get to the point of adding a second vendor, we'd need a second compiled version of shimpei pointing to that second path anyway.

arnaldo2792 replied:
I could make this change, but I would prefer if we don't hard-code the vendor specific name in shimpei. I understand that the contents of the `hook.json` file are vendor specific, but there won't be any case in which we have two "vendor runtimes" enabled in the same AMI, and shimpei doesn't really care about the hook that will be executed, it only knows that there is a `hook.json` file that `oci-add-hooks` will use.

The `hook.json` file can be provided by a `kmod-<kernel>-<vendor>` package, just like I did with `kmod-5.10-nvidia`, and we will save ourselves the N shimpei compilations, per vendor.

bcressey replied:
We could avoid a second copy of `shimpei` by doing the multicall / busybox thing where we inspect arg0 and see if we're called as `nvidia-oci-hook`, `amdgpupro-hook`, etc, and use the matching `/usr/share/oci-add-hooks/<prefix>.json` as our input.

Then packages that wanted to use `shimpei` could just install a symlink, e.g. `/usr/bin/nvidia-oci-hook -> /usr/bin/shimpei`.

samuelkarp commented on packages/libnvidia-container/0004-Use-NVIDIA_PATH-to-look-up-binaries.patch:
```diff
@@ -0,0 +1,36 @@
+         }
+ 
+         while ((dir = strsep(&ptr, ":")) != NULL) {
++                log_infof("Looking at %s", dir);
```
We probably don't actually want to have this log line here.  Once we have libnvidia-container working, this output is just noise.

arnaldo2792 replied:
:sweat_smile: I saw this line after I did my last forced push. I already got rid off it.

bcressey replied:
Is the change that removes this line coming at a later point? I thought I was looking at the current state of the patches.

samuelkarp commented on packages/oci-add-hooks/oci-add-hooks.spec:
```diff
@@ -0,0 +1,59 @@
+
+%build
+%cross_go_configure %{goimport}
+GO111MODULE=off go build -o oci-add-hooks -ldflags "-linkmode=external"
```
```
GO111MODULE=off go build -buildmode=pie -ldflags="-linkmode=external" -o oci-add-hooks .
```

samuelkarp commented on packages/containerd/containerd-config-toml_k8s_nvidia:
```diff
@@ -0,0 +1,39 @@
+# Pause container image is specified here, shares the same image as kubelet's pod-infra-container-image
+sandbox_image = "{{settings.kubernetes.pod-infra-container-image}}"
+
+[plugins."io.containerd.grpc.v1.cri".containerd]
+default_runtime_name = "nvidia"
```
Why are we setting this as the default?  Wouldn't the typical experience be to have it available as an option set through [RuntimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/)?

arnaldo2792 replied:
The typical experience with EKS Optimzed AMI, CO'OS, and the GPU Operator is that the users don't have to set a `RuntimeClass` to run their workloads. This settings helps us have parity with at least the EKS Optimized AMI.

arnaldo2792 commented:
Forced pushed includes:
- Use the same binutils version as the one in the SDK, along with a check to prevent drifts
- Optional `nvidia.env` file in `containerd.service`
- Change dependency order for `oci-add-hooks` and `nvidia-container-toolkit`
- Replaced `/etc` with `%_corss_sysconfigdir` in `glibc.spec`
- Set `IPMI_HANDLER` as a module in the kernel config
- Templated files in `kmod-5.10-nvidia` package
- Leave just the bare minimum NVIDIA libraries for me to test, the libraries will be added in a different PR
- Removed `nvidia-persistenced`, will be added in a different PR
- Load `i2c_core` and `impi_msghandler` at boot with a `modules-load.d` drop-in
- Changed driver's configuration file shape
- Separate linking and loading kernel modules sets in driverdog, and use two subsequent oneshot services
- Add `node-feature-discovery/features.d` to all k8s variants
- Use the SDK's LD linker in `libnvidia-containerd`
- Moved drop-in to delay `ldconfig.service` to `kmod-5.10-nvidia`

bcressey commented on packages/binutils/binutils.spec:
```diff
@@ -0,0 +1,70 @@
+Name: %{_cross_os}binutils
+Version: 2.35.2
+Release: 1%{?dist}
+Summary: A set of GNU tools for working with binaries
```
```suggestion
Summary: Tools for working with binaries
```

bcressey commented on packages/binutils/binutils.spec:
```diff
@@ -0,0 +1,70 @@
+%{summary}.
+
+%package devel
+Summary: Files for development using binutils
```
```suggestion
Summary: Files for development using tools for working with binaries
```

bcressey commented on macros/shared:
```diff
@@ -206,6 +206,7 @@ CROSS_COMPILATION_CONF_EOF\
 %__objcopy %{_bindir}/%{_cross_target}-objcopy
 %__objdump %{_bindir}/%{_cross_target}-objdump
 %__strip %{_bindir}/%{_cross_target}-strip
+%__ld %{_bindir}/%{_cross_target}-ld
```
These override the default Fedora macros, mostly so that debuginfo extraction works properly. Making changes like this is rather brittle since it affects the operation of code that we don't track explicitly.

We shouldn't need to set or use `%__ld` in our own spec files.

bcressey commented on packages/binutils/binutils.spec:
```diff
@@ -0,0 +1,70 @@
+
+%build
+# Fail if the SDK version is different than the one provided in the image
+test %{version} = $(%{__ld} -v | cut -d ' ' -f 5) || exit 1
```
nit: prefer `[` to `test`.

I'd also recommend using `awk` to get the last field rather than using the field count - this one gets the version field for our build and the default `/usr/bin/ld`:

```suggestion
[ %{version} = $(%{_cross_target}-ld -v | | awk '{print $NF}') ] || exit 1
```

bcressey commented on packages/binutils/binutils.spec:
```diff
@@ -0,0 +1,70 @@
+rm -rf %{buildroot}%{_cross_mandir}
+rm -rf %{buildroot}%{_cross_localedir}
+rm -rf %{buildroot}%{_cross_libdir}/ldscripts
+rm -f %{buildroot}%{_cross_libdir}/lib*.la
```
Typically I exclude these in `%files` rather than removing them, with the rationale that it's somewhat easier to add those files later if needed.

bcressey commented on packages/binutils/binutils.spec:
```diff
@@ -0,0 +1,70 @@
+rm -f %{buildroot}%{_cross_libdir}/lib*.la
+
+%files
+%license COPYING
```
This is a copy of the `GPL-2.0-or-later` license. There's a `COPYING3` file too, but it makes me less confident that your choice of `GPL-3.0-or-later` as the overall package license is correct.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+%global kernel_version 5.10.59
```
Need to avoid hard-coding this. We should be able to get it from the extracted kernel archive.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+%{nil}
+
+mkdir -p %{_cross_rootdir}/usr/src/kernels
+tar -xf %{_cross_datadir}/bottlerocket/kernel-devel.tar.xz -C %{_cross_rootdir}/usr/src/kernels
```
Ugh, this shouldn't work and it's a bug in the SDK that it does - the sysroot is owned by `builder` but should be owned by `root`.

You're not supposed to be allowed to write outside of `%_builddir` or `%buildroot` - definitely shouldn't be unpacking stuff to `/`.

bcressey replied:
Opened https://github.com/bottlerocket-os/bottlerocket-sdk/issues/64 to track this issue.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+%global nvidia_tesla_470_version 470.57.02
+%global nvidia_tesla_versions %{nvidia_tesla_470_version}
+%global nvidia_kmod_modules nvidia nvidia-modeset nvidia-uvm nvidia-peermem
+%global kernel_sources %{_cross_rootdir}/usr/src/kernels/kernel-devel
```
I would tend to define these closer to the point of use - putting the version globals at the top makes sense because they can affect the `Version` field.

`%nvidia_kmod_modules` doesn't seem to be used and I'd probably avoid the loop construct that uses `%nvidia_tesla_versions` until we have a second version of that driver.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+
+%package tesla-470
+Summary: NVIDIA 470 tesla driver for the Linux kernel
+Requires: %{name}
```
```suggestion
%package tesla-470
Summary: NVIDIA 470 Tesla driver
Version: %{nvidia_tesla_470_version}
Requires: %{name}
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+  INSTALL_HDR_PATH="%{buildroot}%{_cross_prefix}"\\\
+  INSTALL_MOD_PATH="%{buildroot}%{_cross_prefix}"\\\
+  INSTALL_MOD_STRIP=1\\\
+%{nil}
```
I don't think we actually need this `%kmake` macro; it was an artifact of when we were trying to troubleshoot the build earlier. You could just use `make` below without any of these parameters.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+tar -xf %{_cross_datadir}/bottlerocket/kernel-devel.tar.xz -C %{_cross_rootdir}/usr/src/kernels
+
+%build
+# TODO: this might change for the grid driver
```
If we don't have a really clear idea of how to do this, I'd omit these references. 
```suggestion
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+for v in %{nvidia_tesla_versions}; do
+  pushd NVIDIA-Linux-%{_cross_arch}-$v/kernel
+
+  # We set IGNORE_CC_MISMATCH even though we are using the same compiler used to compile the kernel
```
The comment should explain "why" since the "what" seems obvious 😁

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+  %{__strip} -g --strip-unneeded nvidia/nv-interface.o
+  %{__strip} -g --strip-unneeded nvidia-uvm.o
+  %{__strip} -g --strip-unneeded nvidia-peermem/nvidia-peermem.o
+  %{__strip} -g --strip-unneeded nvidia-modeset/nv-modeset-interface.o
```
I alluded to this in my comment on the macros, but I'd rather use `%{_cross_target}-strip` here - it's possible I'll need to change the `%__strip` definition at some point for debuginfo reasons.

bcressey replied:
Seems like we're missing `nvidia-drm.o` here - at least I had that in my [Dockerfile](https://gist.github.com/bcressey/ed9b70b0ddd4b85bfbf1c343d8a3af3b) build.

Even if we don't need to ship it, I'd rather build it and exclude it in `%files` so it's more clear that we're following with the steps in `yum-packaging-precompiled-kmod`.

It'd also be worth linking to that spec so it's clear where these steps are coming from.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+  done
+
+  # Link our own .o files from the -interface.o and the -kernel.o
+  %{__ld} -r -o nvidia-modeset.o nvidia-modeset/nv-modeset-interface.o nvidia-modeset/nv-modeset-kernel.o
```
should be `%{_cross_target}-ld` and not `%{__ld}`

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+
+  # Link our own .o files from the -interface.o and the -kernel.o
+  %{__ld} -r -o nvidia-modeset.o nvidia-modeset/nv-modeset-interface.o nvidia-modeset/nv-modeset-kernel.o
+  %{__ld} -r -o nvidia-peermem.o nvidia-peermem/nvidia-peermem.o
```
Not sure why we're not linking `nvidia.o` here, and it seems like we're not stripping these objects.

arnaldo2792 replied:
In both specs they don't ship the `nvidia.o` file, they link it at runtime:

https://github.com/NVIDIA/yum-packaging-precompiled-kmod/blob/5def99b662dafbd6d132d5d2aa29a1fe84817c4a/dnf-kmod-nvidia.spec#L230

Also, do I need to strip the objects here, even though the files used to create them were already stripped?

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+%dir %{_cross_libdir}/systemd/system/ldconfig.service.d
+%{_cross_tmpfilesdir}/nvidia.conf
+%{_cross_libdir}/systemd/system/
+%{_cross_libdir}/systemd/system/ldconfig.service.d/ldconfig.conf
```
Considering it more, I think I'd rather put this in `release` where the `systemd-tmpfiles-setup` override is now. Having it there means we can reliably use tmpfiles to adjust cache behavior for other cases where we might need to do that.

The reason I changed my mind is that I did a lot of profiling of `ldconfig` related changes while looking at boot speed - e.g. booting an image with a prebuilt cache - and it didn't make any perceptible difference.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+# Common NVIDIA conf files from 200 to 299
+Source200: nvidia-tmpfiles.conf.in
+Source202: ldconfig-service.conf
+Source203: modules-load.d.conf
```
nit: to match the convention we use for other drop-ins like sysctls:
```suggestion
Source203: nvidia-dependencies-modules-load.conf
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+  # We delete these files since we just stripped the input .o files above
+  for m in nvidia nvidia-modeset nvidia-peermem; do
+    rm ${m}.o
+  done
```
```suggestion
rm nvidia{,-modeset,-peermem}.o
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+
+  # Delete the .ko files created in make command, just to be safe that we
+  # don't include any linked module in the base image
+  rm *.ko
```
I really would like this section to look more like either the "upstream" spec or that Dockerfile adaptation - following all the steps except the final `ko` build.

The blanket removal here of `*.ko` just raises more questions. We won't include any files we don't list in the package manifest.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+done
+
+%install
+
```
nit: extra newline
```suggestion
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+# Begin NVIDIA tesla 470
+pushd NVIDIA-Linux-%{_cross_arch}-%{nvidia_tesla_470_version}
+# We install bins and libs in a versioned directory to prevent collisions with future drivers versions
+install -d %{buildroot}%{_cross_bindir}/nvidia/tesla/%{nvidia_tesla_470_version}/
```
All of these binaries are probably better off in `%{_cross_libexecdir}/nvidia/tesla/bin` rather than a `/usr/bin` type path. 

Having mutually conflicting directories that could end up in `$PATH` is hard to reason about, and `/usr/libexec` is meant for more plugin-style binaries where some other process manages the invocation.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+install -d %{buildroot}%{_cross_libdir}/nvidia/tesla/%{nvidia_tesla_470_version}/
+# We don't install the .o files in /_cross_lib_dir/modules, because they are
+# not kernel modules, and depmod will fail.
+install -d %{buildroot}%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/modules
```
This is sort of an obvious comment to make and like most of these comments, I'd probably either drop it or replace it with a comment on the overall strategy rather than the specific mechanics. I like the comments in the kernel spec file as examples of this - they're less about "what" and more about "why are we doing this crazy thing or using these specific options".

In this case I think we could clarify the non-module nature of these objects by changing the directory name:
```suggestion
# We don't install the .o files in /_cross_lib_dir/modules, because they are
# not kernel modules, and depmod will fail.
install -d %{buildroot}%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+install -m 755 nvidia-cuda-mps-control %{buildroot}%{_cross_bindir}/nvidia/tesla/%{nvidia_tesla_470_version}
+install -m 755 nvidia-cuda-mps-server %{buildroot}%{_cross_bindir}/nvidia/tesla/%{nvidia_tesla_470_version}
+# nvidia-ngx-updater
+install -m 755 nvidia-ngx-updater %{buildroot}%{_cross_bindir}/nvidia/tesla/%{nvidia_tesla_470_version}
```
I'm OK with installing these but might want to `%exclude` them if we're not sure they're useful for containers.

For example it's not clear how helpful `nvidia-cuda-mps-server` will be if we don't actually run it ourselves, and I'm not clear from the name if `nvidia-ngx-updater` is trying to update something on the host.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+%{_cross_attribution_file}
+%dir %{_cross_tmpfilesdir}
+%dir %{_cross_factorydir}
+%dir %{_cross_unitdir}
```
This spec file does not own these directories; it should only own its specific files.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+%{_cross_libdir}/modules-load.d/nvidia-dependencies.conf
+
+%files tesla-470
+%license NVIDIA-Linux-%{_cross_arch}-%{nvidia_tesla_470_version}/LICENSE
```
Is this the only relevant license file? Probably need to consult with the experts on this.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,190 @@
+%{_cross_bindir}/nvidia/tesla/%{nvidia_tesla_470_version}/*
+%{_cross_libdir}/nvidia/tesla/%{nvidia_tesla_470_version}/*
+%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/modules/*
+%{_cross_factorydir}/nvidia/tesla/%{nvidia_tesla_470_version}/*
```
I would prefer to list out all these files because it's pretty important to keep track of what we're actually including. Possibly the libraries can get the wildcard treatment when there are a lot of them.

bcressey commented on packages/kmod-5.10-nvidia/ldconfig-service.conf:
```diff
@@ -0,0 +1,2 @@
+[Unit]
+After=local-fs.target systemd-tmpfiles-setup.service
```
`After` may be an additive setting so you should only need to add `systemd-tmpfiles-setup.service` here.

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tesla-ld.so.conf.in:
```diff
@@ -0,0 +1 @@
+/usr/lib/nvidia/tesla/NVIDIA_VERSION/
```
nit: it'd be somewhat better to use a token like `__LIBDIR__` here that we replace with the sed, to avoid hardcoding paths.

`packages/libbzip2/bzip2.pc.in` is an example of this.

I also don't love the trailing slash, since it seems obvious from context that it's a directory.

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tesla-path.env.in:
```diff
@@ -0,0 +1 @@
+NVIDIA_PATH=/usr/bin/nvidia/tesla/NVIDIA_VERSION/
```
nit: can omit trailing slash, might also benefit from not hardcoding the expected path

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tesla-tmpfiles.conf.in:
```diff
@@ -0,0 +1,4 @@
+C /etc/drivers/driver-config.toml - - - - /usr/share/factory/nvidia/tesla/NVIDIA_VERSION/build-config.toml
+C /etc/containerd/nvidia.env - - - - /usr/share/factory/nvidia/tesla/NVIDIA_VERSION/nvidia-path.env
+C /etc/ld.so.conf.d/nvidia-tesla.ld.so.conf - - - - /usr/share/factory/nvidia/tesla/NVIDIA_VERSION/ld.so.conf
+C /etc/systemd/system/nvidia-persistenced.service.d/nvidia-persistenced.conf - - - - /usr/share/factory/nvidia/tesla/NVIDIA_VERSION/nvidia-persistenced.conf
```
still refers to persistence daemon config

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tesla-tmpfiles.conf.in:
```diff
@@ -0,0 +1,4 @@
+C /etc/drivers/driver-config.toml - - - - /usr/share/factory/nvidia/tesla/NVIDIA_VERSION/build-config.toml
```
This would be somewhat more ergonomic if `driverdog` read all the config files in the directory - then the tmpfile could be in `/usr/share/factory` under its final name (`nvidia-tesla-VERSION.toml`).

bcressey commented on packages/libnvidia-container/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/NVIDIA/libnvidia-container/archive/v1.5.0.tar.gz"
+sha512 = "da33212a7b169a21d7db8821a09ab351604a678d857217ca9de237cf0d0ae4e5a11146fae773276d42c06e1b1a056889275c1a72588004e2cab5528c91c5c1a9"
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/NVIDIA/nvidia-modprobe/archive/450.57/nvidia-modprobe-450.57.tar.gz"
```
For lack of a better place to ask - should we be using `nvidia-modprobe` to load the nvidia drivers rather than just plain `modprobe`?

arnaldo2792 replied:
`libnvidia-container` uses `nvidia-modprobe`'s sources, not the binary itself. According to the documentation, `nvidia-modprobe` will try to load the driver if not already loaded, and create the character device files. Also, from the docs:

> Users should not normally need to run nvidia-modprobe  by  hand:  the  NVIDIA  user-space driver components will fork(2)/exec(3) it when needed.

In my tests, I haven't have to do anything to have the character devices ready, so I think it is safe not to run `nvidia-modprobe` ourselves for now. Also, `driverdog` won't be calling `nvidia-modprobe` to load the drivers, but rather we will have another service to call it.

bcressey commented on packages/libnvidia-container/libnvidia-container.spec:
```diff
@@ -0,0 +1,78 @@
+Version: 1.5.0
+Release: 1%{?dist}
+Summary: NVIDIA container runtime library
+License: Apache-2.0 and GPL-2.0-only
```
Another license to double check. The other licenses may be for vendored dependencies we don't use?

```
❯ askalono crawl .
./COPYING
License: GPL-3.0-only (original text)
Score: 0.998
./COPYING.LESSER
License: LGPL-3.0-only (original text)
Score: 0.998
./NOTICE
Error: Confidence threshold not high enough for any known license
./LICENSE
License: Apache-2.0 (original text)
Score: 0.999
```

It should be aligned with the license files we're including.

bcressey commented on packages/nvidia-container-toolkit/nvidia-container-toolkit.spec:
```diff
@@ -0,0 +1,52 @@
+%license LICENSE
+%{_cross_attribution_file}
+%{_cross_bindir}/nvidia-container-runtime-hook
+%{_cross_datadir}/oci-add-hooks/hook.json
```
Can we give this a more specific name, like `nvidia-container-runtime-hook.json` or `nvidia-hook.json`?

bcressey commented on packages/oci-add-hooks/oci-add-hooks.spec:
```diff
@@ -0,0 +1,51 @@
+
+%build
+%cross_go_configure %{goimport}
+# We use `GO111MODULE=off` to force golang to look for the dependencines in the GOPATH
```
```suggestion
# We use `GO111MODULE=off` to force golang to look for the dependencies in the GOPATH
```

bcressey commented on packages/os/link-kernel-modules.service:
```diff
@@ -0,0 +1,15 @@
+Description=Load additional kernel modules
+RefuseManualStart=true
+RefuseManualStop=true
+After=preconfigured.target
+Wants=preconfigured.target
```
I'm not thrilled about this since we're sort of creating a limbo between `preconfigured.target` and `configured.target` where stuff happens but it's not in either phase.

I'd rather depend specifically on the services in `preconfigured.target` that we need to run after, and turn this into a required service for that target.

bcressey commented on packages/os/load-kernel-modules.service:
```diff
@@ -0,0 +1,15 @@
+Description=Load additional kernel modules
+RefuseManualStart=true
+RefuseManualStop=true
+After=preconfigured.target link-kernel-modules.service
+Wants=preconfigured.target link-kernel-modules.service
```
Same note about dependency ordering - let's keep all of this happening in `preconfigured.target`.

bcressey commented on variants/aws-k8s-1.21-nvidia/Cargo.toml:
```diff
@@ -0,0 +1,41 @@
+    "kernel-5.10",
+    "kubelet-1.21",
+    "release",
+    "binutils",
```
`binutils` should turn into a dependency of `driverdog` since it's something of an implementation detail there. Otherwise this dependency graph looks good to me.

arnaldo2792 committed:
packages: add NVIDIA tesla 470 driver

This commit adds the NVIDIA tesla 470 driver. The drivers are
subpackages of `kmod-5.10-nvidia`, since we don't want to have a spec
file per driver type per driver version. The spec file for
`kmod-5.10-nvidia` holds the drivers that are compatible with the 5.10
kernel. New spec files should be added for newer kernel and driver
versions.

The `kmod-5.10-nvidia` package provides a tmpfilesd conf file to create
the lib modules directory where the kernel modules will be created.

Each subpackage installs the libraries and binaries underneath
`%{_cross_bindir}/nvidia/<type>` and `%{_cross_libdir}/nvidia/<type>`
respectively to prevent collisions while building the subpackages.
Kernel module objects are installed in
`%{_cross_datadir}/nvidia/<type>/modules`, so that `driverdog` can
compile them at runtime.

Each subpackage provides a drop-in configuration file for containerd,
that sets the `NVIDIA_PATH` environment variable. This environment
variable must be set to the directory that contains the NVIDIA userland
tools, which will be mounted on the containers by `libnvidia-container`.
The environment variable is set for containerd, since
`libnvidia-container` is called by the runtime to set up the containers.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Forced push includes:

- Remove variant from github actions
- Fix both bintuils.spec description and licenses
- Add a check to fail if the binutils version is different from the one in the SDK
- Exclude files in bintuils.spec instead of deleting them
- Change shimpei to use the `0` argument as a prefix to look for the `<prefix>-hook.json` file
- Add the `nvidia-oci` symlink to `shimpei` in `nvidia-container-toolkit`
- Update containerd's configuration file to use `nvidia-oci` as the runtime binary
- Update `libnvidia-container`'s licenses
- Use the `ld` and `strip` binaries in the SDK
- Install NVIDIA binaries in `%{_cross_libexecdir}`
- Use templated files in `kmod-5.10-nvidia`
- Build `nvidia-drm` but exclude it in `kmod-5.10-nvidia`
- List all the objects instead of using the `*` in `kmod-5.10-nvidia`
- Delay `ldconfig.service` in `release.spec` instead of `kmod-5.10-nvidia`
- Move `link-kernel-modules` and `load-kernel-modules` to `preconfigured.target`
- `driverdog` now reads all the files in `/etc/drivers` , instead of looking for one specific file
- Add `binutils` as a dependency of `release` but only include it in variants that include `driverdog`

arnaldo2792 commented:
Rebased develop to fix conflicts

bcressey commented on packages/containerd/containerd-config-toml_k8s_nvidia:
```diff
@@ -0,0 +1,39 @@
+[plugins."io.containerd.grpc.v1.cri".registry.mirrors."{{@key}}"]
+endpoint = [{{join_array ", " this }}]
+{{/each}}
+{{/if}}
```
It'd be worth comparing this with the original again; you're missing changes from a64a4c76f at least.

bcressey commented on packages/oci-add-hooks/oci-add-hooks.spec:
```diff
@@ -0,0 +1,51 @@
+%build
+%cross_go_configure %{goimport}
+# We use `GO111MODULE=off` to force golang to look for the dependencies in the GOPATH
+GO111MODULE=off go build -ldflags "-linkmode=external" -o oci-add-hooks
```
Should also set `-buildmode=pie`

bcressey commented on packages/nvidia-container-toolkit/nvidia-container-toolkit.spec:
```diff
@@ -0,0 +1,55 @@
+
+BuildRequires: %{_cross_os}glibc-devel
+Requires: %{_cross_os}libnvidia-container
+Requires: %{_cross_os}oci-add-hooks
```
nit: since `shimpei` depends on `oci-add-hooks`, and this package doesn't invoke it directly, we can remove this dependency to flatten the build graph a bit.

bcressey commented on packages/release/ldconfig-service.conf:
```diff
@@ -0,0 +1,2 @@
+[Unit]
+After=systemd-tmpfiles-setup.service
```
Would be nice to add a comment indicating why we're changing the default ordering.

bcressey commented on sources/shimpei/src/main.rs:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    // Use the `execv` syscall instead of the std::process::Command struct, since
+    // the struct will call `posix_spawn` under the hood.
```
I don't understand this comment - what struct? why does it matter that it calls `posix_spawn`? - but the behavior seems correct in that we want to fully replace this process rather than leaving behind a parent.

bcressey commented on sources/shimpei/src/main.rs:
```diff
@@ -0,0 +1,95 @@
+        CString::new("--hook-config-path").context(error::InvalidString)?,
+        CString::new(hook_path.display().to_string()).context(error::InvalidString)?,
+        CString::new("--runtime-path").context(error::InvalidString)?,
+        CString::new(RUNC_BIN_PATH).context(error::InvalidString)?,
```
For the fixed strings it'd be fine to just `unwrap` or `expect`. The hook path and the additional arguments are less obvious although they can't realistically contain a 0 byte either.

bcressey commented on variants/aws-k8s-1.21-nvidia/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+[package]
+# This is the aws-k8s-1.21 variant.  "." is not allowed in crate names, but we
```
```suggestion
# This is the aws-k8s-1.21-nvidia variant.  "." is not allowed in crate names, but we
```

bcressey commented on variants/aws-k8s-1.21-nvidia/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+    "nvidia-container-toolkit",
+    "kmod-5.10-nvidia-tesla-470"
+]
+kernel-parameters = [
+    "console=tty0",
+    "console=ttyS0",
+]
```
Another good one to compare with the original - it's missing the change from 65daf913e.

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tesla-tmpfiles.conf.in:
```diff
@@ -0,0 +1,3 @@
+C /etc/drivers/nvidia-tesla-NVIDIA_VERSION.toml
```
nit: I'd use the underscores around the replacement variables like `__NVIDIA_X__` consistently, so that the usage in files like `nvidia-tesla-path-env.in` is not ambiguous:
```suggestion
C /etc/drivers/nvidia-tesla-__NVIDIA_VERSION__.toml
```

bcressey reviewed (commented):
Reviewed everything except the kmod packaging and driverdog.

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -0,0 +1,439 @@
+    // Link all the modules sets if no target module was given
+    for driver_config in modules_sets.values() {
+        link_modules(&driver_config, &kernel_version)?;
+    }
```
Should this be in an "else" branch?

arnaldo2792 replied:
Line 114 returns so we shouldn't need the else branch.

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tesla-build-config.toml.in:
```diff
@@ -0,0 +1,18 @@
+lib-modules-path = "kernel/drivers/extra/video/nvidia/tesla"
+objects-source = "NVIDIA_MODULES"
+
+[nvidia-tesla.kernel-modules.nvidia]
+dependencies = ["nvidia.mod.o"]
+
+[nvidia-tesla.kernel-modules.nvidia.runtime-dependencies.nvidia]
+dependencies = ["nv-interface.o", "nv-kernel.o", "nvidia.mod.o"]
+
+[nvidia-tesla.kernel-modules.nvidia-modeset]
+dependencies = ["nvidia-modeset.mod.o"]
+
+[nvidia-tesla.kernel-modules.nvidia-modeset.runtime-dependencies.nvidia-modeset]
+dependencies = ["nv-modeset-interface.o", "nv-modeset-kernel.o"]
+
+[nvidia-tesla.kernel-modules.nvidia-uvm]
+dependencies = ["nvidia-uvm.o", "nvidia-uvm.mod.o"]
```
I find it difficult to wrap my brain around this format, since although I know it translates into a series of `ld` invocations, it's not clear exactly which output files become input files.

I think it'd help to use the "real" file names, even though TOML makes that unlovely. I'd also recommend using the same paths as the dnf kmod packaging, so it's easier to compare with the `%post` scriptlet steps.

Here's an example of what I have in mind:
```
[nvidia-tesla.object-files."nvidia.o"]
link-objects = [
  "nvidia/nv-interface.o",
  "nvidia/nv-kernel.o"
]

[nvidia-tesla.object-files."nvidia-modeset.o"]
link-objects = [
  "nvidia-modeset/nv-modeset-interface.o",
  "nvidia-modeset/nv-modeset-kernel.o",
]

[nvidia-tesla.kernel-modules."nvidia.ko"]
link-objects = [
  "nvidia.o",
  "nvidia.mod.o",
]

[nvidia-tesla.kernel-modules."nvidia-modeset.ko"]
link-objects = [
  "nvidia-modeset.o",
  "nvidia-modeset.mod.o",
]
```

For each driver set, `driverdog` would first work through the set of object files (if any), then it would build the kernel modules.

That also addresses the (hypothetical) case where the same object needs to be built at runtime, and is used by two different kernel modules - the flatter structure allows us to specify the link steps and execute `ld` only once.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,200 @@
+install -m 0644 ld.so.conf %{buildroot}%{_cross_factorydir}/nvidia/tesla/%{nvidia_tesla_470_version}/ld.so.conf
+
+install kernel/nvidia.mod.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d
+install kernel/nvidia/nv-interface.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d
+install kernel/nvidia/nv-kernel.o_binary %{buildroot}%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d/nv-kernel.o
```
nit: add comment above each block of related files like the dnf spec does, e.g.
```suggestion
# driver
install kernel/nvidia/nv-interface.o %{buildroot}%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d
install kernel/nvidia/nv-kernel.o_binary %{buildroot}%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d/nv-kernel.o

# uvm
...
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,200 @@
+%{_cross_factorydir}/nvidia/tesla/%{nvidia_tesla_470_version}/ld.so.conf
+%{_cross_factorydir}/nvidia/tesla/%{nvidia_tesla_470_version}/nvidia-path.env
+
+# Kernel module objects
+%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d/nvidia.mod.o
+%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d/nv-interface.o
+%{_cross_datadir}/nvidia/tesla/%{nvidia_tesla_470_version}/module-objects.d/nv-kernel.o
```
nit: I'd repeat the proposed comment "headers" from the `%install` section for each group of related files

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,200 @@
+# will be build at runtime in the host
+rm nvidia{,-modeset,-peermem}.o
+
+%{_cross_target}-ld -r -o nvidia-modeset.o nvidia-modeset/nv-modeset-interface.o nvidia-modeset/nv-modeset-kernel.o
+%{_cross_target}-ld -r -o nvidia-peermem.o nvidia-peermem/nvidia-peermem.o
```
The first `ld` invocation here seems useless, since we don't actually package the resulting file. The second one is creating a relocatable version of `nvidia-peermem.o` which we then install. This seems like a subtle side effect in the dnf spec, and I'm not entirely sure it's required.

In the interests of clarity, I'd remove both these lines.

If we end up needing the `ld -r` invocation for `nvidia-peermem.o`, I'd add an "object file" step so `driverdog` can generate it from `nvidia-peermem/nvidia-peermem.o`, like we do for `nvidia-modeset.o`.

arnaldo2792 commented:
Forced push includes:

- Fixed containerd's configuration file
- Removed unused linked files from nvidia-5.10-kmod.spec file
- Simplified tmpfilesd conf file to copy `ld.so.conf.d` files
- Added underscores to templated variables
- Refactored driverdog to simplify build configuration file
- Removed binutils and nvidia-container-toolkit from specs where they aren't needed
- Added missing build flag in oci-add-hooks spec file
- Use `expect` in shimpei constant strings

arnaldo2792 commented:
Forced push to fix conflicts + include commits created from this PR, in other PRs.

arnaldo2792 commented:
Forced push includes:

- Fixed OCI hooks API for NVIDIA variant
- Rebased driverdog PR

arnaldo2792 committed:
actions: add matrix `fetch-upstream` variable

The `fetch-upstream` variable is be used to fetch upstream sources when
they aren't provided in the lookaside cache.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Last two pushes include configurations to run github workflows for the nvidia variant. For this variant, we need to fallback to fetch some sources from the upstream instead of the lookaside cache.

arnaldo2792 commented:
Forced push to fix conflicts + drop commits included in other PRs

arnaldo2792 commented:
Forced push includes:

- Edit github actions for NVIDIA variant
- Add missing URL for NVIDIA tesla driver
- Removed NVIDIA License
- Use `bottlerocket-licenses-tool` in kmod-5.10-nvidia
- Cleaned extra lines added in previous rebase

arnaldo2792 commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,210 @@
+Release: 1%{?dist}
+Summary: NVIDIA drivers for the 5.10 kernel
+# We use these licences because we only ship our own software in the main package,
+# each subpackage includes the LICENSE file provided by the run scripts
```
I need to fix/remove this line, to explain that subpackages include the license provided by the `Licenses.toml` configuration

arnaldo2792 commented:
Forced push to rebase changes in libnvidia-container + k8s-device plugin

arnaldo2792 commented:
Forced push to fix build for aarch64 + add k8s device plugin to nvidia variant

bcressey commented on .github/workflows/build.yml:
```diff
@@ -27,27 +27,55 @@ jobs:
           - variant: aws-dev
             arch: x86_64
             supported: false
+            fetch-upstream: "false"
```
Do we have to override this for each item in the list, even if the default isn't changing?

arnaldo2792 replied:
I could try removing it and see if it works :+1:

arnaldo2792 replied:
It didn't :smile: , I changed it back per our offline conversation.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,214 @@
+%global nvidia_tesla_470_version 470.82.01
+%global spdx_id %(bottlerocket-license-tool -l $PWD/rpmbuild/BUILD/Licenses.toml spdx-id nvidia)
+%global license_file %(bottlerocket-license-tool -l $PWD/rpmbuild/BUILD/Licenses.toml path nvidia -p ./licenses)
```
```suggestion
%global spdx_id %(bottlerocket-license-tool -l %{_builddir}/Licenses.toml spdx-id nvidia)
%global license_file %(bottlerocket-license-tool -l %{_builddir}/Licenses.toml path nvidia -p ./licenses)
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,214 @@
+License: Apache-2.0 OR MIT
+URL: http://www.nvidia.com/
+
+# NVIDIA .run scripts from 0 to 199
+Source0: https://us.download.nvidia.com/tesla/%{nvidia_tesla_470_version}/NVIDIA-Linux-%{_cross_arch}-%{nvidia_tesla_470_version}.run
```
nit: it is considered bad form to conditionally include files like this in the spec file since then a source RPM won't be complete - whether it includes the x86 or arm64 run file will depend on what it was built for rather than what the end user of the srpm wants to build for.

In the context of our distro, it's fine since we don't build source RPMs, but I'd still prefer to include both `.run` files and conditionally invoke the right one, just as a stylistic twitch.

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,214 @@
+tar -xf %{_cross_datadir}/bottlerocket/kernel-devel.tar.xz
+
+%build
+pushd NVIDIA-Linux-%{_cross_arch}-%{nvidia_tesla_470_version}/kernel
```
nit: we could remove these pushd & popd calls, and potentially fold the subpackage into the main package based on today's conversation, since it seems like we'll only ship the "production" Tesla drivers.

bcressey replied:
on second thought - let's leave this scaffolding in place in case we end up wanting to ship multiple versions.

bcressey commented on packages/kmod-5.10-nvidia/nvidia-tmpfiles.conf.in:
```diff
@@ -0,0 +1 @@
+D /lib/modules/KERNEL_VERSION/kernel/drivers/extra/video/nvidia/tesla 0755 root root -
```
nit:
```suggestion
D /lib/modules/__KERNEL_VERSION__/kernel/drivers/extra/video/nvidia/tesla 0755 root root -
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,214 @@
+install -d %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}/{drivers,ld.so.conf.d}
+
+KERNEL_VERSION=$(cat %{kernel_sources}/include/config/kernel.release)
+sed -e "s|KERNEL_VERSION|$KERNEL_VERSION|" %{S:200} > nvidia.conf
```
obligatory nit:
```suggestion
sed -e "s|KERNEL_VERSION|${KERNEL_VERSION}|" %{S:200} > nvidia.conf
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,214 @@
+install -m 755 nvidia-debugdump %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{nvidia_tesla_470_version}
+install -m 755 nvidia-cuda-mps-control %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{nvidia_tesla_470_version}
+install -m 755 nvidia-cuda-mps-server %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{nvidia_tesla_470_version}
+%if "%{_cross_arch}" == "x86_64"
+install -m 755 nvidia-ngx-updater %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{nvidia_tesla_470_version}
+%endif
```
This doesn't exist for the arm64 driver?

arnaldo2792 replied:
No :disappointed: , the `.run` archive doesn't include it :disappointed:

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,214 @@
+install -m 755 nvidia-ngx-updater %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{nvidia_tesla_470_version}
+%endif
+
+# TODO: add remaining libraries once we implement "image per variant" in the rpm2img script
```
```suggestion
# TODO: add remaining libraries
```

bcressey commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -0,0 +1,214 @@
+# TODO: add remaining libraries once we implement "image per variant" in the rpm2img script
+# misc
+# Add libnvidia-ml.so for testing purposes
+install  -m755 libnvidia-ml.so.%{nvidia_tesla_470_version} %{buildroot}%{_cross_libdir}/nvidia/tesla/%{nvidia_tesla_470_version}
```
nit:
```suggestion
install -m755 libnvidia-ml.so.%{nvidia_tesla_470_version} %{buildroot}%{_cross_libdir}/nvidia/tesla/%{nvidia_tesla_470_version}
```

arnaldo2792 commented:
Forced push includes documentation changes

arnaldo2792 commented:
Forced push to rebase the upstream

bcressey commented on BUILDING.md:
````diff
@@ -121,6 +121,34 @@ licenses = [
+
+If you want to build the `aws-k8s-1.21-nvidia` variant, you can follow these steps to prepare a `Licenses.toml` file using the [License for customer use of NVIDIA software](https://www.nvidia.com/en-us/drivers/nvidia-license/):
+
+1. Fetch the drivers `.run` archive from the URL provided in the [kmod-5.10-nvidia package](packages/kmod-5.10-nvidia/Cargo.toml), i. e.:
+
+```shell
+curl -LO https://us.download.nvidia.com/tesla/470.82.01/NVIDIA-Linux-x86_64-470.82.01.run
+```
+
+2. Extract the sources and copy the `LICENSE` file to the `licenses` directory in your Bottlerocket root directory:
````
It seems cleaner just to document using the HTML license file; then we won't have driver links that go out of date.

bcressey commented on QUICKSTART-EKS.md:
```diff
@@ -158,11 +158,11 @@ aws ssm get-parameters --region us-west-2 \
+Also be aware that when operating in GovCloud the IAM ARNs will need to be updated to the following: `arn:aws-us-gov`.
+For example:
+ `arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy`
+ will be updated to:
```
nit: I like the cleanup but I tend to want it in a separate commit - can still be in this PR though

bcressey commented on QUICKSTART-EKS.md:
```diff
@@ -369,3 +369,12 @@ Once it launches, you should be able to run pods on your Bottlerocket instance u
+
+The `aws-k8s-1.21-nvidia` variant includes the required packages and configurations to leverage NVIDIA GPUs.
+It comes with the [NVIDIA Tesla driver](https://docs.nvidia.com/datacenter/tesla/drivers/index.html) along with the libraries required by the [CUDA toolkit](https://developer.nvidia.com/cuda-toolkit) included in your orchestrated containers.
+It also includes the [NVIDIA k8s device plugin](https://github.com/NVIDIA/k8s-device-plugin), so please make sure you don't have a daemonset running the device plugin in your cluster before launching a new node using this variant.
```
```suggestion
It also includes the [NVIDIA k8s device plugin](https://github.com/NVIDIA/k8s-device-plugin).
If you already have a daemonset for the device plugin in your cluster, you may need to use taints and tolerations to keep it from running on Bottlerocket nodes.
```

bcressey commented on QUICKSTART-EKS.md:
```diff
@@ -369,3 +369,12 @@ Once it launches, you should be able to run pods on your Bottlerocket instance u
+It comes with the [NVIDIA Tesla driver](https://docs.nvidia.com/datacenter/tesla/drivers/index.html) along with the libraries required by the [CUDA toolkit](https://developer.nvidia.com/cuda-toolkit) included in your orchestrated containers.
+It also includes the [NVIDIA k8s device plugin](https://github.com/NVIDIA/k8s-device-plugin), so please make sure you don't have a daemonset running the device plugin in your cluster before launching a new node using this variant.
+
+With this variant, most of the existing NVIDIA tools (like [DCGM](https://github.com/NVIDIA/dcgm-exporter) and the [GPU Feature Discovery](https://github.com/NVIDIA/gpu-feature-discovery)) work as you would expect, you can install them in your cluster following the instructions provided for each project.
```
```suggestion
Additional NVIDIA tools such as [DCGM](https://github.com/NVIDIA/dcgm-exporter) and [GPU Feature Discovery](https://github.com/NVIDIA/gpu-feature-discovery) will work as expected.
You can install them in your cluster by following the `helm install` instructions provided for each project.
```

bcressey commented on QUICKSTART-EKS.md:
```diff
@@ -369,3 +369,12 @@ Once it launches, you should be able to run pods on your Bottlerocket instance u
+It also includes the [NVIDIA k8s device plugin](https://github.com/NVIDIA/k8s-device-plugin), so please make sure you don't have a daemonset running the device plugin in your cluster before launching a new node using this variant.
+
+With this variant, most of the existing NVIDIA tools (like [DCGM](https://github.com/NVIDIA/dcgm-exporter) and the [GPU Feature Discovery](https://github.com/NVIDIA/gpu-feature-discovery)) work as you would expect, you can install them in your cluster following the instructions provided for each project.
+Even though you could use the [GPU Operator](https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/getting-started.html#install-nvidia-gpu-operator) to install these tools, we recommend installing these tools individually via `helm install` since the Operator could break your workloads on upgrades.
```
This is hard to understand without more details. I'd suggest saying something like:
```suggestion
The [GPU Operator](https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/getting-started.html#install-nvidia-gpu-operator) can also be used to install these tools.
However, it is cumbersome to select the right subset of features to avoid conflicts with the software included in the variant.
Therefore we recommend installing the tools individually if they are required.
```

bcressey commented on README.md:
```diff
@@ -54,6 +54,9 @@ The following variants support EKS, as described above:
 - `aws-k8s-1.21`
+- `aws-k8s-1.21-nvidia`
+
+Please refer to [this document](QUICKSTART-EKS.md#aws-k8s-121-nvidia-variant) to learn more about the `aws-k8s-1.21-nvidia` variant.
```
I'd prefer not to link to variant specifics here, or else we'll be doing that everywhere on this page. In general, we need to figure out a more structured approach to getting started with particular variants, since this isn't scaling very well.

bcressey commented on SECURITY_FEATURES.md:
```diff
@@ -134,6 +134,8 @@ All binaries are linked with the following options:
 
 Together these enable [full RELRO support](https://www.redhat.com/en/blog/hardening-elf-binaries-using-relocation-read-only-relro) which makes [ROP](https://en.wikipedia.org/wiki/Return-oriented_programming) attacks more difficult to execute.
 
+**Note:** the `aws-k8s-1.21-nvidia` variant includes the NVIDIA k8s device plugin which is compiled without the `-wl,-z,now` flags, and other precompiled NVIDIA libraries that could haven't been compiled with hardening flags.
```
```suggestion
**Note:** Certain variants, such as the ones for NVIDIA, include precompiled binaries that may not have been built with these hardening flags.
```

arnaldo2792 committed:
packages: change %_topdir rpm macro

There are some rpm commands in the Dockerfile executed as 'root', which
results in macros like `%{_builddir}` to point to `/root` instead of
`/home/builder`. By changing `%_topdir` to `/home/builder`, macros that
rely on this value will always point to the build directory.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Forced push uses `%{_builddir}` macro in `kmod-5.10-nvidia` spec

bcressey commented on BUILDING.md:
````diff
@@ -121,6 +121,34 @@ licenses = [
+cargo make fetch-licenses -e BUILDSYS_UPSTREAM_LICENSES_FETCH=true
+```
+
+3. Build your image, setting the `BUILDSYS_UPSTREAM_SOURCE_FALLBACK` flag to `true`, if you haven't cache the driver's sources:
````
```suggestion
3. Build your image, setting the `BUILDSYS_UPSTREAM_SOURCE_FALLBACK` flag to `true`, if you haven't cached the driver's sources:
```

arnaldo2792 commented:
Forced push fixes:

- KERNEL_VERSION to __KERNEL_VERSION__
- Type in BUILDING.md

jpculp commented on .github/workflows/build.yml:
```diff
@@ -27,30 +27,59 @@ jobs:
         variant: [aws-k8s-1.18, aws-k8s-1.19, aws-k8s-1.20, aws-k8s-1.21, aws-ecs-1]
         arch: [x86_64, aarch64]
         supported: [true]
+        fetch-upstream: ["false"]
```
nit. our previous booleans (and values) were unquoted, but these ones are quoted. I think I would unquote these for consistency.

arnaldo2792 replied:
The previous booleans values were used within the context of the github workflows, which are literal booleans. Since GH worklflows are defined in YAML, I can't use `false` and expect it to get translated to `"false"`, I could get a `1`, or another value that doesn't work for the environment variables passed to `cargo make`

jpculp reviewed (approved):
We should include a `releases-url`, but not necessarily this PR.

zmrow reviewed (approved):
:night_with_stars:
